vochain: implements deterministic process ID type

p4u opened the pull request:
Introduce a new type `ProcessID` with Marshal and Unmarshal methods.

The new identifier holds information about the process
- chainID: a 6 bytes trunked hash of the blockchain identifier
- organization: hex address of the wallet creating the process
- census proof type: kind of census used for the process (from protobuf models)
- envelope type: flags for configuring the ballot counting mode
- nonce: an incremental uint32 number

So processID: 395b41cc9abcd8da6bf26964af9d7eed9e03e53415d37aa96045062100000001  represents
- chID: vocdoni-azeno
- addr: 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045
- proof: arbo
- envelopetype: 0x21
- nonce: 1

This commit only introduces the new type, not the required changes on the blockchain logic.

Discussion: https://github.com/vocdoni/protocol/discussions/28

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: implements deterministic process ID type

Introduce a new type `ProcessID` with Marshal and Unmarshal methods.

The new identifier holds information about the process
- chainID: a 6 bytes trunked hash of the blockchain identifier
- organization: hex address of the wallet creating the process
- census origin: kind of census used for the process (from protobuf models)
- envelope type: flags for configuring the ballot counting mode
- nonce: an incremental uint32 number

So processID: b9180c3de924d8da6bf26964af9d7eed9e03e53415d37aa96045061300000002  represents
- chID: vocdoni-azeno
- addr: 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045
- csOrigin: 0x06
- envType: 10011 (0x13)
- nonce: 2

This commit only introduces the new type, not the required changes on the blockchain logic.

Discussion: https://github.com/vocdoni/protocol/discussions/28

Signed-off-by: p4u <pau@dabax.net>

jordipainan reviewed (approved):
LGTM!

mvdan commented on vochain/process_id.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// Marshal encodes to bytes
+func (p *ProcessID) Marshal() []byte {
```
might as well implement https://pkg.go.dev/encoding#BinaryMarshaler and its unmarshaler counterpart

mvdan commented on vochain/process_id.go:
```diff
@@ -0,0 +1,120 @@
+// chID:vocdoni-azeno addr:0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045 pt:arbo nc:1
+type ProcessID struct {
+	chainID          string
+	organizationAddr *common.Address
```
I don't think you should be using a pointer here. a pointer weighs 8 bytes usually; an address is just 20 bytes.

mvdan commented on vochain/process_id.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// Addr returns the organizer address
+func (p *ProcessID) Addr() *common.Address {
```
per the above, I don't think this should return a pointer (SetAddr is already good)

mvdan commented on vochain/process_id.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+// Proof returns a string representing the census proof type
+func (p *ProcessID) Proof() string {
```
nit: SetProof takes a pointer, but Proof takes a string. would be nice for the two to be consistent.

p4u committed:
vochain: use the deterministic process ID approach

With some exceptions in order to keep backwards compatibility.
While standard accounts are forced to use the new format,
oracles are allowed to introduce arbitrary processIDs.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
vochain: update genesis and make staticcheck happy

Merge branch 'master' into deterministic_processID'

Signed-off-by: p4u <pau@dabax.net>
